drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/base_loader.rb

```diff
@@ -0,0 +1,135 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HmisExternalApis::AcHmis::Importers::Loaders
+  class BaseLoader
+    attr_reader :reader, :clobber, :tracker, :table_names
+
+    def self.perform(...)
+      new(...).perform
+    end
+
+    def initialize(reader:, tracker: nil, clobber:)
+      @reader = reader
+      @clobber = clobber
+      @tracker = tracker
+      @table_names = []
+    end
+
+    protected
+
+    def supports_upsert?
+      false
+    end
+
+    def runnable?
+      clobber ? true : supports_upsert?
+    end
+
+    # 'YYYY-MM-DD HH24:MM:SS'
+    DATE_TIME_FMT = '%Y-%m-%d %H:%M:%S'.freeze
+    def valid_date?(str)
+      str =~ /\A\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\z/
+    end
+
+    def parse_date(str)
+      return unless str
+      raise ArgumentError, "Invalid date-time format. Expected 'YYYY-MM-DD HH24:MM:SS' but got '#{str}'" unless valid_date?(str)
+
+      DateTime.strptime(str, DATE_TIME_FMT)
+    end
+
+    def cde_definition(owner_type:, key:)
+      @cache ||= {}
+      @cache[[owner_type, key]] ||= cde_definitions.find_or_create(owner_type: owner_type, key: key)
+    end
+
+    def cde_definitions
+      @cde_definitions ||= CustomDataElementDefinitions.new(data_source_id: data_source.id, system_user_id: system_user_id)
+    end
+
+    def row_value(row, field:, required: true)
+      value = row[field]&.strip&.presence
+      raise "field '#{field}' is missing" if required && !value
```

Comment:
When this error comes up, will we know which file it's coming from? Hopefully we can see it in the stack trace, just want to confirm. Maybe we should log the whole row so we can more easily find the culprit too.

Reply:
Good idea, I'll add the full row to the error message here.

So the idea is that CsvFile decorates any exception that occurs as the file is being read (using rows.each) with the relevant filename and row number. That should let us locate the problem row. Does that seem sufficient?

Reply:
Definitely! thanks